File: leetcode/146.lru-cache.py
# @leet start
from collections import OrderedDict


class LRUCache:
    def __init__(self, capacity: int):
        # Map stores two_tuples
        self.map: OrderedDict[int, int] = OrderedDict()
        self.capacity: int = capacity

    def get(self, key: int) -> int:
        if key in self.map:
            self.map.move_to_end(key)

            return self.map[key]

        return -1

    def put(self, key: int, value: int) -> None:
        is_update = key in self.map

        if len(self.map) == self.capacity and not is_update:
            self.map.popitem(last=False)

        self.map[key] = value

        if is_update:
            self.map.move_to_end(key)


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)
# @leet end

File: leetcode/146.lru-cache.ts
// @leet start
class MyNode {
  constructor(
    public key: number,
    public val: number,
    public prev: MyNode | null = null,
    public next: MyNode | null = null,
  ) {}
}

class LRUCache {
  private size = 0;
  private capacity: number;
  private keyToNode = new Map<number, MyNode>();
  private head = new MyNode(-1, -1); // sentinel
  private tail = this.head;

  constructor(capacity: number) {
    this.capacity = capacity;
  }

  private moveToTail(nd: MyNode) {
    if (nd === this.tail) {
      return;
    }

    nd.prev!.next = nd.next;
    nd.next!.prev = nd.prev;

    nd.next = null;
    nd.prev = this.tail;
    this.tail.next = nd;
    this.tail = this.tail.next;
  }

  private addToTail(key: number, value: number) {
    this.tail.next = new MyNode(key, value, this.tail);
    this.tail = this.tail.next;

    this.keyToNode.set(key, this.tail);
  }

  // Precondition: size >= 1
  private removeOldest() {
    const toDelete = this.head.next!;
    this.head.next = toDelete.next;

    if (this.size === 1) {
      this.tail = this.head;
    } else {
      toDelete.next!.prev = this.head;
    }

    this.keyToNode.delete(toDelete.key);
    this.size--;
  }

  get(key: number): number {
    const target = this.keyToNode.get(key);

    if (!target) {
      return -1;
    }

    this.moveToTail(target);
    return target.val;
  }

  put(key: number, value: number) {
    const target = this.keyToNode.get(key);

    if (target != null) {
      target.val = value;
      this.moveToTail(target);
    } else {
      if (this.size == this.capacity) {
        this.removeOldest();
      }

      this.size++;
      this.addToTail(key, value);
    }
  }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * var obj = new LRUCache(capacity)
 * var param_1 = obj.get(key)
 * obj.put(key,value)
 */
// @leet end
